### ISSUE TITLE
Local auth should not require an external URL to be set

### ISSUE THREAD

**Issue by jeffawang** [concourse/concourse#2722]
## What challenge are you facing?

While using the web frontend with local auth, clicking the "login" button redirects to either a single configured external URL or 127.0.0.1, but I don't know beforehand what URL should be configured. In my particular case, users are accessing concourse through multiple different URLs, so there is no single correct value for the URL.

In our situation, users on the local network are able to use internal DNS to access the machine (eg. `http://concourse.internal:8080`), while users in another network access it through an ssh tunnel (eg. `http://localhost:8080`). This stems from DNS that is unresolvable or IPs that are not routable in different parts of the network. Off the top of my head, this could result from firewalls, access restrictions, network partitions, split-horizon DNS, etc. We don't have a VPN or want to maintain one right now.

Furthermore, when concourse is running behind a reverse proxy such as apache or nginx, imho the external URL should be within the purview of that frontend web server, and concourse should not mind if it changes. In this case, if I wanted to change the DNS/base URL of that frontend web server, I believe that I shouldn't also have to reconfigure the external URL for concourse.

Issue #2463 is related in that it is DNS-related. That issue dealt with DNS being unresolvable from within concourse's network, while this issue points to a broader scope that also includes user DNS resolution.

For now, I'm stuck on 3.14.1, where all local auth works and my use cases are supported.

## What would make this better?

I would be delighted if every time I was able to access the concourse web instance, I was able to log in.

## Are you interested in implementing this yourself?

Sorry no, I don't have the time or concourse-implementation-specific expertise right now!

**Comment by aepsil0n** [comment]
This is also a bit of a roadbump when trying out Concourse locally for first-time users. After checking out and starting the docker-compose configuration for Concourse, I intuitively went to https://localhost:8080. The login does not work though, as `localhost` is not `127.0.0.1`, I guess? Anyway, it took me a while to figure that out. It makes the login functionality seem a bit broken, which does not give a great first impression.

**Comment by eedwards-sk** [comment]
If your external users reverse proxy in with the same port, could you create a hosts entry that maps to the expected hostname of concourse?

e.g. if your external url is `http://concourse.internal:8080`. then when a user connects through reverse-proxy, add a hosts entry with `127.0.0.1 concourse.internal`

**Comment by jeffawang** [comment]
@eedwards-sk Thanks for mentioning that workaround. It's true, that could solve it when users are going through the reverse proxy.

Consider this facet though: some users are on laptops and could start being in the local network temporarily (ie. where `concourse.internal` is resolvable). They would have to modify their hosts file whenever they transition in or out of the network, or they would have to keep running the proxy and, depending on routing rules, potentially go out to the Internet and back.

**Comment by eedwards-sk** [comment]
@jeffawang As a recommendation, based on personal experience with concourse, it sounds like you might be stretching a complex network architecture around it, rather than building the appropriate network architecture to support it.

The issue you referenced that got fixed was because concourse was incorrectly using the external url internally. It sounds like you inadvertently ended up with a use case that that was relying on that behavior, or it might just be the auth change in 4.x

If you want a more seamless experience and you're unwilling to set up a VPN, concourse runs great behind a public load balancer and you can put oauth on it, just like https://ci.concourse-ci.org/ does.

**Comment by eedwards-sk** [comment]
Also, you can probably automate some aspects of changing the hosts file when setting up an ssh tunnel, but that's getting way outside appropriate scope for a github issue here.

**Comment by jeffawang** [comment]
@eedwards-sk, you're right. And your workarounds are theoretically valid -- any way to make the external url the only url would work.

However, I think this discussion should be directed away from my particular network infrastructure and towards the fact that my initial expectations were not met. 

As mentioned in the original post, "I would be delighted if every time I was able to access the concourse web instance, I was able to log in." @aepsil0n's localhost vs 127.0.0.1 example is another demonstration of this expectation not being met.

If this is unfeasible or I'm being unreasonable, then some straightforward indication that this is out of scope from the project maintainers would be fine. I would adjust my expectations. But right now it seems to me like a regression.

The workaround I've chosen so far (the past few months) is to stay on the pi version (3.14.1) and keep an eye on this issue.

**Comment by xtremerui** [comment]
@jeffwang curious what version of concourse are you on now? Is this still an issue for you?
